### ISSUE TITLE
Fold with string array

### ISSUE THREAD

**Issue by nu8** [rust-lang/rust#76792]
I tried some code like this:

~~~
fn main() {
   let a = vec!["May", "June"];
   let s = a.iter().fold("", |s2, s3|
      s2 + s3
   );
   println!("{}", s == "MayJune");
}
~~~

Result:

~~~
error[E0369]: cannot add `&&str` to `&str`
 --> a.rs:4:10
  |
4 |       s2 + s3
  |       -- ^ -- &&str
  |       |  |
  |       |  `+` cannot be used to concatenate two `&str` strings
  |       &str
  |
help: `to_owned()` can be used to create an owned `String` from a string
reference. String concatenation appends the string on the right to the string
on the left and may require reallocation. This requires ownership of the string
on the left
  |
4 |       s2.to_owned() + s3
  |       ^^^^^^^^^^^^^
~~~

Ok, fair enough. So I change my code to exactly that. But then I get this:

~~~
error[E0308]: mismatched types
 --> a.rs:4:7
  |
4 |       s2.to_owned() + s3
  |       ^^^^^^^^^^^^^^^^^^
  |       |
  |       expected `&str`, found struct `std::string::String`
  |       help: consider borrowing here: `&(s2.to_owned() + s3)`
~~~

Ok, fair enough. So I change my code to exactly that. But then I get this:

~~~
error[E0515]: cannot return reference to temporary value
 --> a.rs:4:7
  |
4 |       &(s2.to_owned() + s3)
  |       ^--------------------
  |       ||
  |       |temporary value created here
  |       returns a reference to data owned by the current function
~~~

Why is Rust giving bogus suggestion, and is what I am trying to do possible?
Note, I would prefer to avoid suggestions such as "just use `join`" or similar,
as this question is intended to address a more generic problem. Rust version:

~~~
rustc 1.46.0 (04488afe3 2020-08-24)
~~~

I posted this on Stack Overflow https://stackoverflow.com/q/63912961, and
someone suggested the source of the error is the init value, so this would work
instead:

~~~rust
fn main() {
   let a = vec!["May", "June"];
   let s = a.iter().fold(String::new(), |s2, s3|
      s2 + s3
   );
   println!("{}", s == "MayJune");
}
~~~

It was also suggested that I post this here, in case anyone wants to review the
`rustc` help system. One person described it like this:

> despite the best intentions, because the compiler's view is too local it
> guilelessly leads you down a completely useless path of frustration.

**Comment by leonardo-m** [comment]
Using a fold for this purpose isn't a good idea. It's better to use:
https://doc.rust-lang.org/nightly/std/primitive.slice.html#method.concat

I don't know if Clippy gives this hint.

**Comment by nu8** [comment]
@leonardo-m the reason for this post isnt the specific example given, or how best to solve the specific example given. Its more to shine some light on how in some cases the Rust help system can actually be hurtful rather than helpful.

It might be worth discussing if the Rust system is being too eager to offer help, especially in cases where its "confidence level" is low.

**Comment by steveklabnik** [comment]
> It might be worth discussing if the Rust system is being too eager to offer help, especially in cases where its "confidence level" is low.

Discussions are better served on internals.rust-lang.org than the issue tracker; that being said, this specific scenario might be considered a bug, so I'll leave this open for now.

**Comment by nu8** [comment]
@the8472 exactly. In my mind, its a conversation like this between the user and the compiler:

> User: compile this
>
> Compiler: wont work, what about this?
>
> U: well, will what youre suggesting actually compile?
>
> C: No
>
> U: then why are you suggesting it?
>
> C: ...

**Comment by the8472** [comment]
@nu8 the compiler may be suggesting a first step towards a fix. In many cases incomplete code can require more than one change to compile.
